### PR TITLE
`azurerm_windows_function_app`, `azurerm_linux_function_app` - fix content share not being created automatically issue

### LINKED ISSUE #19134: azurerm_linux_function_app Storage options not working
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

1.2.0

### AzureRM Provider Version

3.29.1

### Affected Resource(s)/Data Source(s)

azurerm_linux_function_app

### Terraform Configuration Files

```hcl
resource "azurerm_linux_function_app" "function_apps" {
  name                          = "myapp"
  resource_group_name           = azurerm_resource_group.main.name
  location                      = var.location
  storage_account_name          = azurerm_storage_account.function_app_storage.name
  storage_account_access_key    = azurerm_storage_account.function_app_storage.primary_access_key
  service_plan_id               = azurerm_service_plan.asp.id
  https_only                    = true

  identity {
    type         = "UserAssigned"
    identity_ids = [azurerm_user_assigned_identity.function_apps.id]
  }

  site_config {
    application_insights_connection_string = azurerm_application_insights.function_app.connection_string
    application_insights_key               = azurerm_application_insights.function_app.instrumentation_key

    application_stack {
      python_version = "3.9"
    }

    cors {
      allowed_origins = [
        "mydomain.com"
      ]
      support_credentials = true
    }
  }

  app_settings = {
    APP_PARTITION_KEY            = var.app_partition_key
    ENVIRONMENT                  = var.environment
    HOME_URL                     = var.home_url
  }
}
```
```


### Debug Output/Panic Output

```shell
-
```

### Expected Behaviour

  storage_account_name and storage_account_access_key are set, and the promised vars should have been generated. They have not. 

From the docs:
"Note: For storage related settings, please use related properties that are available such as storage_account_access_key, terraform will assign the value to keys such as WEBSITE_CONTENTAZUREFILECONNECTIONSTRING, AzureWebJobsStorage in app_setting."

AzureWebJobsStorage vars are set, but WEBSITE_CONTENTAZUREFILECONNECTIONSTRING and WEBSITE_CONTENTSHARE are not.

When creating through the Portal, this all works. The docs are either wrong or incomplete, or the module is not doing what it should.

### Actual Behaviour

The environment variables that were promised, are not generated. This causes deployments to fail, because there is no storage.
Screen Shot 2022-11-03 at 19 41 16

### Steps to Reproduce

_No response_

### Important Factoids

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by xiaxyi**
Fix content share not being created automatically if `WEBSITE_CONTENTOVERVNET` sets to 1.

According to the official [guidance](https://docs.microsoft.com/en-us/azure/azure-functions/functions-app-settings#website_contentshare), _"There are scenarios where you must set the WEBSITE_CONTENTSHARE value to a predefined share, such as when you [use a secured storage account in a virtual network](https://docs.microsoft.com/en-us/azure/azure-functions/configure-networking-how-to#restrict-your-storage-account-to-a-virtual-network). In this case, you must set a unique share name for the main function app and the app for each deployment slot."_
If the property `WEBSITE_CONTENTOVERVNET` set to 1 in `app_setting` while the users don't specify or specify a non-existed storage file share, they get the error message:

```
reading logs configuration for Windows Function App: (Site Name "xxx" / Resource Group "xxx"): web.AppsClient#GetDiagnosticLogsConfiguration: Failure responding to request: StatusCode=503 -- Original Error: autorest/azure: Service returned an error. Status=503 Code="ServiceUnavailable" Message="The operation timed out and could not be completed. Please retry the action or try again later." Details=[{"Message":"The operation timed out and could not be completed. Please retry the action or try again later."},{"Code":"ServiceUnavailable"},{"ErrorEntity":{"Code":"ServiceUnavailable","ExtendedCode":"51007","Message":"The operation timed out and could not be completed. Please retry the action or try again later.","MessageTemplate":"The operation timed out and could not be completed. Please retry the action or try again later."}}]
│
```

fix https://github.com/hashicorp/terraform-provider-azurerm/issues/17254
fix https://github.com/hashicorp/terraform-provider-azurerm/issues/14167
fix #19134

### CODE REVIEW

File: internal/services/appservice/helpers/service_plan.go [hunk 1]
```diff
@@ -137,7 +137,7 @@ func PlanTypeFromSku(input string) string {
 }
 
 // ServicePlanInfoForApp returns the OS type and Service Plan SKU for a given App Service Resource
-func ServicePlanInfoForApp(ctx context.Context, metadata sdk.ResourceMetaData, id interface{}) (osType *string, planSku *string, err error) {
+func ServicePlanInfoForApp(ctx context.Context, metadata sdk.ResourceMetaData, id interface{}, serviceFarmId string) (osType *string, planSku *string, err error) {
```
Comment (jackofallops) [inline internal/services/appservice/helpers/service_plan.go:140]: This function is intended to retrieve the Service Plan ID from the App ID, it should not take the ID directly to circumvent this.

File: internal/services/appservice/linux_function_app_resource.go [hunk 2]
```diff
@@ -703,19 +767,21 @@ func (r LinuxFunctionAppResource) Update() sdk.ResourceFunc {
...
 
-			_, planSKU, err := helpers.ServicePlanInfoForApp(ctx, metadata, *id)
+			// Some service plan updates are allowed - see customiseDiff for exceptions
+			var serviceFarmId string
+			if metadata.ResourceData.HasChange("service_plan_id") {
+				serviceFarmId = state.ServicePlanId
+				existing.SiteProperties.ServerFarmID = utils.String(serviceFarmId)
+			}
+
+			_, planSKU, err := helpers.ServicePlanInfoForApp(ctx, metadata, *id, serviceFarmId)
```
Comment (jackofallops) [inline internal/services/appservice/linux_function_app_resource.go:777]: This should be a call to the new service plan to get the SKU, and not repurpose the helper function.

File: internal/services/appservice/windows_function_app_resource.go [hunk 3]
```diff
@@ -536,6 +547,22 @@ func (r WindowsFunctionAppResource) Read() sdk.ResourceFunc {
...
+				servicePlan, err := servicePlanClient.Get(ctx, servicePlanId.ResourceGroup, servicePlanId.ServerfarmName)
+				if err != nil {
+					return fmt.Errorf("reading %s: %+v", servicePlanId, err)
+				}
+				if sku := servicePlan.Sku; sku != nil && sku.Name != nil {
+					planSku = sku.Name
+				}
+			}
+			sendContentSettings := helpers.PlanIsElastic(planSku)
+
```
Comment (jackofallops) [inline internal/services/appservice/windows_function_app_resource.go:565]: As above, this functionality shouldn't be present in the Read.

File: internal/services/appservice/windows_function_app_resource.go [hunk 4]
```diff
@@ -546,6 +573,43 @@ func (r WindowsFunctionAppResource) Read() sdk.ResourceFunc {
...
+					exists, err := saFileShareClient.Exists(ctx, account.ResourceGroup, storageAccountName, storageFileShare)
+					if err != nil {
+						return fmt.Errorf("checking for existence of Storage Share %q (Account %q) : %+v", storageFileShare, storageAccountName, err)
+					}
+					if exists == nil {
+						return fmt.Errorf("the storage file share %s for function app %s does not exists, please specify an existing one", storageFileShare, id.ID())
+					}
+				}
+			}
+
```
Comment (jackofallops) [inline internal/services/appservice/windows_function_app_resource.go:612]: As above, this shouldn't be in the Read.

File: internal/services/appservice/windows_function_app_resource.go [hunk 5]
```diff
@@ -702,7 +766,13 @@ func (r WindowsFunctionAppResource) Update() sdk.ResourceFunc {
...
 			}
 
-			_, planSKU, err := helpers.ServicePlanInfoForApp(ctx, metadata, *id)
+			var serviceFarmId string
+			if metadata.ResourceData.HasChange("service_plan_id") {
+				serviceFarmId = state.ServicePlanId
+				existing.SiteProperties.ServerFarmID = utils.String(serviceFarmId)
+			}
+
+			_, planSKU, err := helpers.ServicePlanInfoForApp(ctx, metadata, *id, serviceFarmId)
```
Comment (jackofallops) [inline internal/services/appservice/windows_function_app_resource.go:775]: As above, the helper should not be repurposed, and the SKU should be read from the new service plan ID. A new helper for that would be acceptable.

File: internal/services/appservice/linux_function_app_resource.go [hunk 6]
```diff
@@ -547,6 +575,43 @@ func (r LinuxFunctionAppResource) Read() sdk.ResourceFunc {
...
+					exists, err := saFileShareClient.Exists(ctx, account.ResourceGroup, storageAccountName, storageFileShare)
+					if err != nil {
+						return fmt.Errorf("checking for existence of Storage Share %q (Account %q) : %+v", storageFileShare, storageAccountName, err)
+					}
+					if exists == nil {
+						return fmt.Errorf("the storage file share %s for function app %s does not exists, please specify an existing one", storageFileShare, id.ID())
+					}
+				}
+			}
+
```
Comment (jackofallops) [inline internal/services/appservice/linux_function_app_resource.go:614]: As above, none of this should be in a read function, we should be setting received data back into state only, not performing configuration checks on external resources that we're being told about.

**Review by jackofallops — CHANGES_REQUESTED** [review]
Hi @xiaxyi - Thanks for this PR.

I've left some comments inline below that will need to be addressed before we can continue review.  I think this PR should probably have been split into smaller specific PRs to address the disparate concerns as the changes do not fully match the description / title, which makes the changes difficult to review and will unfortunately slow the process. That said, if you can make the changes here we can get this moving.

Thanks!

File: internal/services/appservice/windows_function_app_resource.go (hunk 4 above)
Comment (xiaxyi) [reply internal/services/appservice/windows_function_app_resource.go:612]: Thanks for the review, this is to address the timeout issue. if the user doesn't specify a predefined file share and also set the WEBSITE_CONTENTOVERVNET in the app_setting, the provider will timeout by trying to access a non-existence file share. I can add the 4.0 condition to this check block.

File: internal/services/appservice/linux_function_app_resource.go (hunk 6 above)
Comment (xiaxyi) [reply internal/services/appservice/linux_function_app_resource.go:614]: the same to the reply in the windows function app

File: internal/services/appservice/windows_function_app_resource.go (hunk 3 above)
Comment (xiaxyi) [reply internal/services/appservice/windows_function_app_resource.go:565]: We need to get the service plan id to determine if the function app is elastic or not, then to check if the website_content_share property.

**Comment by xiaxyi** [comment]
> Hi @xiaxyi - Thanks for this PR.
> 
> I've left some comments inline below that will need to be addressed before we can continue review. I think this PR should probably have been split into smaller specific PRs to address the disparate concerns as the changes do not fully match the description / title, which makes the changes difficult to review and will unfortunately slow the process. That said, if you can make the changes here we can get this moving.
> 
> Thanks!

Thanks @jackofallops for the feedbacks. All the code changes in this pr is tend to resolve one issue which is the content share issue under elastic plan.

As I mentioned in the description, if user sets `WEBSITE_CONTENTOVERVNET ` without specifying a predefined storage file share, there will be error accessing that share. 

The file share check in read function is to ensure the predefined file share exists, otherwise, we throw error instead of waiting for the read function to time out. 

The service plan check is to determine whether the function is elastic plan.

File: internal/services/appservice/windows_function_app_resource.go (hunk 4 above)
Comment (tombuildsstuff) [reply internal/services/appservice/windows_function_app_resource.go:612]: agreed - this shouldn't be in the Read function - we don't retrieve the file share in the Read, so this is unnecessary

Comment (xiaxyi) [reply internal/services/appservice/windows_function_app_resource.go:612]: Thanks @tombuildsstuff for the comments. If we don't check the storage file share, shall we just let the read function timeout?

Comment (tombuildsstuff) [reply internal/services/appservice/windows_function_app_resource.go:612]: @xiaxyi why will the read function timeout? it's calling the ARM API and interacting with the Storage API at all?

Comment (xiaxyi) [reply internal/services/appservice/windows_function_app_resource.go:612]: The function app is trying to connect to the file share specified in the `[WEBSITE_CONTENTSHARE](https://learn.microsoft.com/en-us/azure/azure-functions/functions-app-settings)`, Generally, the file share is created by function app automatically, however, the it cannot be created if user specify the key `WEBSITE_CONTENTOVERVNET` in `app_setting, it requires the user to create a valid file share in advance and put the value in `WEBSITE_CONTENTSHARE`. We didn't add such notes to user in the current provider, so the function(Read function) times out.

If we don't check the file share in read function, then what's left is to add notes in docs and let the read function timeout.

**Comment by xiaxyi** [comment]
@jackofallops @tombuildsstuff , updated th read function, removed the storage file share check.

**Review by jackofallops — CHANGES_REQUESTED** [review]
Hi @xiaxyi - apologies for the long overdue re-review on this. I hope you don't mind, but I pushed some changes to simplify the updates and remove the need for re-processing some data.  Could you look to add a test to cover the case being addressed here, and I think we'll be good to merge once that's in and passing.

Thanks

**Comment by xiaxyi** [comment]
Thanks @jackofallops for the update, I updated the code and added the test cases for both linux and windows.

feel free to let me know if there is anything else needed.

**Review by jackofallops — CHANGES_REQUESTED** [review]
Hi @xiaxyi 
Thanks for adding the tests. I've re-pushed the changes that were lost during the merge main. The ignores in the test suggest this needs some more work, possibly in the functions I've called out in line below if you can take a look?

Thanks!

**Comment by xiaxyi** [comment]
Thanks @jackofallops for the comments, I applied your change and pushed again, apologize for my mistakes last time. 

Regarding the acc test import steps, I think we still need to keep the ignore step for those app setting:
````
"app_settings.WEBSITE_CONTENTSHARE", "app_settings.WEBSITE_CONTENTAZUREFILECONNECTIONSTRING", "app_settings.%"
```

As I mentioned in my previous comments, service will try to add the `WEBSITE_CONTENTSHARE` and `WEBSITE_CONTENTAZUREFILECONNECTIONSTRING` for the function app instance (in elastic or consumption plan) automatically if no `WEBSITE_CONTENTOVERVNET` is added by user, which means, we need to determine whether to set those kvp back to state by checking the user's TF config and that's why we need to ignore the changes when doing the import. So I'm not going to change the original unpackappsetting function:https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/appservice/linux_function_app_resource.go#L1090

**Comment by jackofallops** [comment]
> Thanks @jackofallops for the comments, I applied your change and pushed again, apologize for my mistakes last time.
> 
> Regarding the acc test import steps, I think we still need to keep the ignore step for those app setting:
> 
> ```
> "app_settings.WEBSITE_CONTENTSHARE", "app_settings.WEBSITE_CONTENTAZUREFILECONNECTIONSTRING", "app_settings.%"
> ```
> 
> As I mentioned in my previous comments, service will try to add the `WEBSITE_CONTENTSHARE` and `WEBSITE_CONTENTAZUREFILECONNECTIONSTRING` for the function app instance (in elastic or consumption plan) automatically if no `WEBSITE_CONTENTOVERVNET` is added by user, which means, we need to determine whether to set those kvp back to state by checking the user's TF config and that's why we need to ignore the changes when doing the import. So I'm not going to change the original unpackappsetting function:[`main`/internal/services/appservice/linux_function_app_resource.go#L1090](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/appservice/linux_function_app_resource.go?rgh-link-date=2023-03-15T02%3A34%3A14Z#L1090)

Hi @xiaxyi - Apologies, you are correct. These _should_ be ignored on Import as the service does not manage them if absent. I'll message you offline to discuss.

**Comment by jackofallops** [comment]
Tests look good

image

**Review by jackofallops — APPROVED** [review]
Thanks @xiaxyi for the updates and patience. This LGTM now 👍

**Pull request merged**